starfish/codebook.py

```diff
@@ -356,13 +358,65 @@ def to_json(self, filename: str) -> None:
         with open(filename, 'w') as f:
             json.dump(code_array, f)
 
-    def decode_euclidean(self, intensities: IntensityTable) -> IntensityTable:
+    @staticmethod
+    def _unit_normalize(array: xr.DataArray) -> xr.DataArray:
+        """unit normalize each feature of array"""
+        total_intensities = array.sum(dim=[Indices.CH.value, Indices.HYB.value])
+        array = array / total_intensities
+
+        # if a code is all zero, the unit intensity should be spread across the channel
+        partitioned_intensity = 1 / np.product(array.shape[1:])
+        array = array.fillna(partitioned_intensity)
+        return array
+
+    @staticmethod
+    def _approximate_nearest_code(
+            norm_codes: xr.DataArray, norm_intensities: xr.DataArray
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """find the nearest code for each feature using the ball_tree approximate NN algorithm
+
+        Parameters
+        ----------
+        norm_codes : Codebook
+            codebook with each code normalized to unit length (sum = 1)
+        norm_intensities : IntensityTable
+            intensity table with each feature normalized to unit length (sum = 1)
+
+        Returns
+        -------
+        np.ndarray : qualities
+            the quality of each feature's match to its closest code (equal to 1 - distance)
+        np.ndarray : gene_ids
+            the gene that corresponds to each matched code
+
+        """
+        linear_codes = norm_codes.values.reshape(norm_codes.shape[0], -1)
+        linear_features = norm_intensities.values.reshape(norm_intensities.shape[0], -1)
+        # reshape into traces
+        nn = NearestNeighbors(n_neighbors=1, algorithm='ball_tree').fit(linear_codes)
+        distances, indices = nn.kneighbors(linear_features)
+        gene_ids = np.ravel(norm_codes.indexes[Codebook.Constants.GENE.value].values[indices])
+        qualities = 1 - np.ravel(distances)
+
+        return qualities, gene_ids
+
+    def decode_euclidean(
```

Comment:
I still think it's weird to have codebooks do the decoding. Is the plan to move this into a Decoder in a later PR, along with the other decode method? I also think the helper functions this uses (e.g., approximate_nearest_code, unit_normalize, etc.) should move out of the decoder as well.

Reply:
Yep, that's still the plan. I made #323 but figured MERFISH was more important to prioritize. There's quite a bit of tech debt that I need to work through from the switch over to xarray based objects.